File: Python/MergeSort.py
def mergeSort(arr):
	if len(arr) > 1:

		# Finding the mid of the array
		mid = len(arr)//2

		# Dividing the array elements
		L = arr[:mid]

		# into 2 halves
		R = arr[mid:]

		# Sorting the first half
		mergeSort(L)

		# Sorting the second half
		mergeSort(R)

		i = j = k = 0

		# Copy data to temp arrays L[] and R[]
		while i < len(L) and j < len(R):
			if L[i] < R[j]:
				arr[k] = L[i]
				i += 1
			else:
				arr[k] = R[j]
				j += 1
			k += 1

		# Checking if any element was left
		while i < len(L):
			arr[k] = L[i]
			i += 1
			k += 1

		while j < len(R):
			arr[k] = R[j]
			j += 1
			k += 1




def printList(arr):
	for i in range(len(arr)):
		print(arr[i], end=" ")
	print()



if __name__ == '__main__':
	
	x = list(map(int,input("Enter the list of numbers separated by space: ").split()))
	print("Given array is", end="\n")
	printList(x)
	mergeSort(x)
	print("Sorted array is: ", end="\n")
	printList(x)

# This code is contributed by Yasas Sandeepa
'''
Input:  x = 1 5 20 3 12 4
Output: x = 1 3 4 5 12 20

Average Time complexity: O(nlogn)
Space complexity: O(n)
'''

File: C/merge_sort.c
#include <stdio.h>
#include <stdlib.h>


void merge(int arr[], int l, int m, int r){
	int i, j, k;
	int n1 = m - l + 1;
	int n2 = r - m;
	int L[n1], R[n2];


	for (i = 0; i < n1; i++)
		L[i] = arr[l + i];
	for (j = 0; j < n2; j++)
		R[j] = arr[m + 1 + j];

	/* Merge the temp arrays back into arr[l..r]*/
	i = 0; // Initial index of first subarray
	j = 0; // Initial index of second subarray
	k = l; // Initial index of merged subarray
	while (i < n1 && j < n2) {
		if (L[i] <= R[j]) {
			arr[k] = L[i];
			i++;
		}
		else {
			arr[k] = R[j];
			j++;
		}
		k++;
	}

	/* Copy the remaining elements of L[], if there
	are any */
	while (i < n1) {
		arr[k] = L[i];
		i++;
		k++;
	}

	/* Copy the remaining elements of R[], if there
	are any */
	while (j < n2) {
		arr[k] = R[j];
		j++;
		k++;
	}
}


void mergeSort(int arr[], int l, int r){
	if (l < r) {
		// Same as (l+r)/2, but avoids overflow for
		// large l and h
		int m = l + (r - l) / 2;

		// Sort first and second halves
		mergeSort(arr, l, m);
		mergeSort(arr, m + 1, r);

		merge(arr, l, m, r);
	}
}


void printArray(int A[], int size){
	int i;
	for (i = 0; i < size; i++)
		printf("%d ", A[i]);
	printf("\n");
}


int main(){

	    // Getting the size of the array from the user
        int n;
        printf("Enter the count of elements in the array: ");
        scanf("%d", &n);

        // Getting the elements of the array from the user
        int arr[n];
        printf("Enter the elements of the array (enter one element and press enter key): ");
        for (int i = 0; i < n; i++) scanf("%d", &arr[i]);

	//int arr[] = { 12, 11, 13, 5, 6, 7 };
	int arr_size = n;

	printf("Given array is \n");
	printArray(arr, arr_size);

	mergeSort(arr, 0, arr_size - 1);

	printf("\nSorted array is \n");
	printArray(arr, arr_size);
	return 0;
}


/*
Merge sort implementation in C
This code is contributed by Yasas Sandeepa
-------------------------------------------
Enter the count of elements in the array: 8
Enter the elements of the array (enter one element and press enter key): 10
4
20
1
7
5
15
12
Given array is 
10 4 20 1 7 5 15 12 

Sorted array is 
1 4 5 7 10 12 15 20 


Complexity Analysis
---------------------
Average Time complexity: O(nlogn)
Space complexity: O(n)
*/
